### PR TITLE
rgw: s3select: add nullif function

### PR DESCRIPTION
**Pull request by albin-antony**
Signed-off-by: Albin Antony <aantony@redhat.com>

### CODE REVIEW

**Comment by galsalomon66** [comment]
due to other merge (not-operator and memory management fixes)
need to resolve conflicts( rebase )

File: include/s3select_functions.h [hunk 1]
```diff
@@ -941,6 +943,38 @@ struct _fn_upper : public base_function {
...
+          throw base_s3select_exception("nullif accept only 2 arguments");
+        }
+        base_statement *first = *iter;
+        x = first->eval();
+        iter++;
+        base_statement *second = *iter;
+        y = second->eval();
+        if (x.type != y.type)
+        {
+          throw base_s3select_exception("type should be same");
```
Comment (galsalomon66) [inline include/s3select_functions.h:967]: its better to be less strict as for numbers
1 = 1.0 (float , integer)

Comment (galsalomon66) [reply include/s3select_functions.h:967]: missing null handling

File: include/s3select_functions.h [hunk 2]
```diff
@@ -941,6 +943,46 @@ struct _fn_upper : public base_function {
...
+        if (args_size != 2)
+        {
+          throw base_s3select_exception("nullif accept only 2 arguments");
+        }
+        base_statement *first = *iter;
+        x = first->eval();
+        iter++;
+        base_statement *second = *iter;
+        y = second->eval();
+        if (x.type == value::value_En_t::NULL || y.type == value::value_En_t::NULL)
```
Comment (galsalomon66) [inline include/s3select_functions.h:965]: instead of x.type == value::value_En_t::NULL 
x.is_null()

File: include/s3select_functions.h [hunk 3]
```diff
@@ -941,6 +943,46 @@ struct _fn_upper : public base_function {
...
+        {
+          result = null;
```
Comment (galsalomon66) [inline include/s3select_functions.h:967]: result is type of pointer to variable , so = null (C++ word) may no fit here

Comment (galsalomon66) [reply include/s3select_functions.h:967]: nullif return null in the case both (left and right) are equal

File: include/s3select_functions.h [hunk 4]
```diff
@@ -941,6 +943,46 @@ struct _fn_upper : public base_function {
...
+          return true;
+        }
+        if ((x.type != value::value_En_t::DECIMAL && x.type != value::value_En_t::FLOAT) || (y.type != value::value_En_t::DECIMAL && y.type != value::value_En_t::FLOAT))
```
Comment (galsalomon66) [inline include/s3select_functions.h:970]: it better to use is_number , is_string , it less error prone, and readable.

File: include/s3select.h [hunk 5]
```diff
@@ -536,6 +536,23 @@ struct push_debug_1 : public base_action
...
 static push_debug_1 g_push_debug_1;
 
+struct push_null : public base_action 
+{
+  void operator()(const char* a, const char* b) const
+  {
+    std::string token(a, b);
+    variable* v;
+
+    if (token.compare("null") == 0 || token.compare("NULL")==0)
```
Comment (galsalomon66) [inline include/s3select.h:546]: push_null is called upon syntax-parser is identifying "null" or "NULL"
do, why to compare again?

File: include/s3select.h [hunk 6]
```diff
@@ -783,11 +800,13 @@ struct s3select : public bsc::grammar<s3select>
 
       log_op = bsc::str_p("and") | bsc::str_p("or");
 
+      null_var = bsc::str_p("null") | bsc::str_p("NULL");
```
Comment (galsalomon66) [inline include/s3select.h:803]: rule is OK is identifying "null"/"NULL"

File: include/s3select.h [hunk 7]
```diff
@@ -763,7 +780,7 @@ struct s3select : public bsc::grammar<s3select>
 
       arithmetic_argument = (float_number)[BOOST_BIND_ACTION(push_float_number)] |  (number)[BOOST_BIND_ACTION(push_number)] | (column_pos)[BOOST_BIND_ACTION(push_column_pos)] |
                             (string)[BOOST_BIND_ACTION(push_string)] |
-                            (function)[BOOST_BIND_ACTION(push_debug_1)]  | (variable)[BOOST_BIND_ACTION(push_variable)] ;//function is pushed by right-term
+                            (function)[BOOST_BIND_ACTION(push_debug_1)]  | (variable)[BOOST_BIND_ACTION(push_variable)] | (null_var)[BOOST_BIND_ACTION(push_null)];//function is pushed by right-term
```
Comment (galsalomon66) [inline include/s3select.h:783]: action is called upon rule is accepted

File: include/s3select_functions.h [hunk 8]
```diff
@@ -941,6 +943,46 @@ struct _fn_upper : public base_function {
...
+        if (args_size != 2)
+        {
+          throw base_s3select_exception("nullif accept only 2 arguments");
+        }
+        base_statement *first = *iter;
+        x = first->eval();
+        iter++;
+        base_statement *second = *iter;
+        y = second->eval();
+        if (x.is_null() || y.is_null())
```
Comment (galsalomon66) [inline include/s3select_functions.h:965]: not correct
only if both are NULL

File: include/s3select_oper.h [hunk 9]
```diff
@@ -567,6 +573,14 @@ class value
     return *this;
   }
 
+  value& operator=(bool b)
+  {
```
Comment (galsalomon66) [inline include/s3select_oper.h:577]: not correct, its override existing functionality. 
who calls this operations?

what should be done , when push_null is called , a variable should be set to null(its value), explicitly.

File: test/s3select_test.cpp [hunk 10]
```diff
@@ -629,3 +629,84 @@ TEST(TestS3selectFunctions, modzero)
     ASSERT_EQ(s3select_result, std::string("0,\n"));
   }
 
+  TEST(TestS3selectFunctions, null)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select 1 + null from stdin;";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:635]: as discussed , this currently could not work properly, need a change in syntax-semantic parsing.
add unit-test for operators.
such as 
TEST(TestS3SElect, float_compare_operator)
TEST(TestS3SElect, arithmetic_operator)

File: test/s3select_test.cpp [hunk 11]
```diff
@@ -629,3 +629,84 @@ TEST(TestS3selectFunctions, modzero)
...
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("nan,\n"));
+}
+
+TEST(TestS3selectFunctions, nullcompare)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select 2 > null from stdin;";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:655]: select  x < y 
 will not accept by syntax-parser

File: include/s3select_oper.h [hunk 12]
```diff
@@ -519,6 +506,10 @@ class value
     return type == value_En_t::S3NULL;
   }
 
+  bool is_na() const
```
Comment (galsalomon66) [inline include/s3select_oper.h:509]: is_nan()

File: include/s3select_functions.h [hunk 13]
```diff
@@ -972,6 +974,50 @@ struct _fn_upper : public base_function {
...
+        x = first->eval();
+        iter++;
+        base_statement *second = *iter;
+        y = second->eval();
+        if (x.is_null() && y.is_null())
+        {
+          result->set_null();
+          return true;
+        }
+        if (x.is_null())
```
Comment (galsalomon66) [inline include/s3select_functions.h:1001]: not correct, null as first argument is legit.

File: include/s3select_functions.h [hunk 14]
```diff
@@ -972,6 +974,50 @@ struct _fn_upper : public base_function {
...
+        {
+          throw base_s3select_exception("Error:first argument should be known");
+        }
+        if (x.is_number() || y.is_number())
+        {
+          if (x.type != y.type)
```
Comment (galsalomon66) [inline include/s3select_functions.h:1007]: if one is float and the second is integer, its ok.
( 1.0 , 1 ) --> return null

File: test/s3select_test.cpp [hunk 15]
```diff
@@ -629,3 +630,130 @@ TEST(TestS3selectFunctions, modzero)
...
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string(",\n"));
```
Comment (galsalomon66) [inline test/s3select_test.cpp:730]: in case of NULL, should print-out NULL
there is a difference between NULL and empty-string

**Comment by galsalomon66** [comment]
the logical operators (AND OR) should changed also. to handle the case where operand = NULL

File: test/s3select_test.cpp [hunk 16]
```diff
@@ -629,3 +630,130 @@ TEST(TestS3selectFunctions, modzero)
...
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("nan,\n"));
```
Comment (mattbenjamin) [inline test/s3select_test.cpp:750]: is the result of arith + null defined to be nan in SQL standards?  I thought the result of such an operation was null.

Comment (galsalomon66) [reply test/s3select_test.cpp:750]: [here](https://docs.google.com/document/d/1edRKxCcbxqNNyH-CsHt3s870eiBdMMjFZknbSVhkOOo/edit?usp=sharing) , it was note that arithmetic operation with NULL return NAN
the thought behind that, is the fact that NAN is supported in C/C++ as for operations (arithmetic and compare) with NAN

however, we could define other rule, such as return NULL for those operations , which mean unknown.
currently, i'm not sure what is the sql-standard for that.

Comment (mattbenjamin) [reply test/s3select_test.cpp:750]: I guess ISO/IEC 9075;  It's of course, not free to download.

Comment (galsalomon66) [reply test/s3select_test.cpp:750]: ok, will review that.
it means that 
1+null -> null 
1+nan -> nan (built-in c/c++) 

1 + null + nan -> null (?) 
nan + 1 + null -> nan (?)

need to decide which (null/nan) got precedence , and stick to that.

Comment (mattbenjamin) [reply test/s3select_test.cpp:750]: I suspect:  any expression involving nan, including 0 or more null terms, results in NaN

File: include/s3select_oper.h [hunk 17]
```diff
@@ -508,7 +508,10 @@ class value
 
   bool is_nan() const
   {
-    return type == value_En_t::S3NAN;
+    if (type == value_En_t::FLOAT) {
```
Comment (galsalomon66) [inline include/s3select_oper.h:511]: ok

File: include/s3select_oper.h [hunk 18]
```diff
@@ -569,6 +625,10 @@ class value
 
   bool operator<(const value& v)//basic compare operator , most itensive runtime operation
   {
+    if ((is_null() || v.is_null()) || (is_nan() || v.is_nan()))
```
Comment (galsalomon66) [inline include/s3select_oper.h:628]: as for optimization,
the nan/null cases are rare compare to others, 
thus, its possible to "push them" down, and save some instructions per comparison.

File: include/s3select_oper.h [hunk 19]
```diff
@@ -701,21 +769,41 @@ class value
...
+      return !(*this == v);
+      } 
   }
-
+  
   template<typename binop> //conversion rules for arithmetical binary operations
   value& compute(value& l, const value& r) //left should be this, it contain the result
   {
     binop __op;
+    if ((l.is_null() || r.is_null()) || (l.is_nan() || r.is_nan()))
```
Comment (galsalomon66) [inline include/s3select_oper.h:801]: optimization issue.
it possible to validate null/nan condition,  upon other conditions (less rare) are false.
and save some instructions per binary-operation

File: include/s3select_oper.h [hunk 20]
```diff
@@ -973,6 +1061,28 @@ class variable : public base_statement
...
+    }
+    else if (reserve_word == s3select_reserved_word::reserve_word_en_t::S3S_NAN)
+    {
+      m_var_type = variable::var_t::COL_VALUE;
+      column_pos = -1;
+      var_value.type = value::value_En_t::S3NAN;
+    }
+    else 
+    {
+      _name = "#";
```
Comment (galsalomon66) [inline include/s3select_oper.h:1080]: ok

File: include/s3select_oper.h [hunk 21]
```diff
@@ -1268,28 +1383,40 @@ class logical_operand : public base_statement
     return std::string("#");//TBD
   }
   virtual value& eval()
-  {bool res;
+  {
+    bool res;
```
Comment (galsalomon66) [inline include/s3select_oper.h:1387]: optimization issue.
in case left side is false , can return false , without evaluate the right side.

File: test/s3select_test.cpp [hunk 22]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
 }
 
+TEST(TestS3SElect, intnull_compare_operator)
+{
+  value a10(10), b11(11), c10(10), d, e;
+  d.type = value::value_En_t::S3NULL;
+  e.type = value::value_En_t::S3NULL;
+  ASSERT_EQ( d > b11, false );
+  ASSERT_EQ( d >= c10, false );
+  ASSERT_EQ( a10 >= c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:232]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 23]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+  ASSERT_EQ( d < a10, false );
+  ASSERT_EQ( d <= b11, false );
+  ASSERT_EQ( d != a10, true );
+  ASSERT_EQ( d != e, true );
+  ASSERT_EQ( d == a10, false );
+  ASSERT_EQ( a10 == c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:238]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 24]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+}
+
+TEST(TestS3SElect, floatnull_compare_operator)
+{
+  value a10(10.1), b11(11.2), c10(10.1), d, e;
+  d.type = value::value_En_t::S3NULL;
+  e.type = value::value_En_t::S3NULL;
+  ASSERT_EQ( d > b11, false );
+  ASSERT_EQ( d >= c10, false );
+  ASSERT_EQ( a10 >= c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:248]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 25]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+  ASSERT_EQ( d < a10, false );
+  ASSERT_EQ( d <= b11, false );
+  ASSERT_EQ( d != a10, true );
+  ASSERT_EQ( d != e, true );
+  ASSERT_EQ( d == a10, false );
+  ASSERT_EQ( a10 == c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:254]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 26]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+}
+
+TEST(TestS3SElect, intnan_compare_operator)
+{
+  value a10(10), b11(11), c10(10), d, e;
+  d.type = value::value_En_t::S3NAN;
+  e.type = value::value_En_t::S3NAN;
+  ASSERT_EQ( d > b11, false );
+  ASSERT_EQ( d >= c10, false );
+  ASSERT_EQ( a10 >= c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:264]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 27]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+  ASSERT_EQ( d < a10, false );
+  ASSERT_EQ( d <= b11, false );
+  ASSERT_EQ( d != a10, true );
+  ASSERT_EQ( d != e, true );
+  ASSERT_EQ( d == a10, false );
+  ASSERT_EQ( a10 == c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:270]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 28]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+}
+
+TEST(TestS3SElect, floatnan_compare_operator)
+{
+  value a10(10.1), b11(11.2), c10(10.1), d, e;
+  d.type = value::value_En_t::S3NAN;
+  e.type = value::value_En_t::S3NAN;
+  ASSERT_EQ( d > b11, false );
+  ASSERT_EQ( d >= c10, false );
+  ASSERT_EQ( a10 >= c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:280]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 29]
```diff
@@ -222,6 +222,70 @@ TEST(TestS3SElect, arithmetic_operator)
...
+  ASSERT_EQ( d < a10, false );
+  ASSERT_EQ( d <= b11, false );
+  ASSERT_EQ( d != a10, true );
+  ASSERT_EQ( d != e, true );
+  ASSERT_EQ( d == a10, false );
+  ASSERT_EQ( a10 == c10, true );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:286]: no need for that, other unit test cover this.

File: test/s3select_test.cpp [hunk 30]
```diff
@@ -629,3 +693,147 @@ TEST(TestS3selectFunctions, modzero)
...
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("nan,\n"));
+} 
+
+TEST(TestS3selectFunctions, nullnanarithmetic)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select 3 * null + nan/5 - 4 * null from stdin;";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:819]: instead of complex query containing several operations.

add test per arithmetic operation (+,/,*-) , with nan/null.
meaning simple binary operation , so we can be sure that all operations are covered.

as in compare operator , write those tests on value object directly.

File: include/s3select_functions.h [hunk 31]
```diff
@@ -972,6 +974,51 @@ struct _fn_upper : public base_function {
...
+          return true;
+        }
+        if (x.is_null())
+        {
+          result->set_null();
+          return true;
+        }
+        if (!(x.is_number() && y.is_number()))
+        {
+          if (x.type != y.type)
```
Comment (galsalomon66) [inline include/s3select_functions.h:1008]: according to [documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-glacier-select-sql-reference-conditional.html#s3-glacier-select-sql-reference-nullif)
there is no error in case of types are not equal.
for one example nullif(5,null) should return 5.

review the examples in the attach AWS documentation.
in other SQL engines (such as oracle), the describe behavior is different. 
currently, we implement AWS behavior.

File: include/s3select_oper.h [hunk 32]
```diff
@@ -345,17 +383,17 @@ struct binop_mult
...
-};
+};   
 
 struct binop_div
 {
   double operator()(double a, double b)
   {
-    if(b == 0)
-    {
+    if (b == 0) {
```
Comment (galsalomon66) [inline include/s3select_oper.h:392]: in case a=nan the result should be nan.

File: include/s3select_oper.h [hunk 33]
```diff
@@ -1268,28 +1403,50 @@ class logical_operand : public base_statement
     return std::string("#");//TBD
   }
   virtual value& eval()
-  {bool res;
+  {
+    bool res;
+    value a = l->eval();
+    value b = r->eval();
```
Comment (galsalomon66) [inline include/s3select_oper.h:1409]: it's not optimized.
before AND/OR actual calculation, both side are evaluated. 
only the left side should be evaluated.

File: test/s3select_test.cpp [hunk 34]
```diff
@@ -222,6 +222,86 @@ TEST(TestS3SElect, arithmetic_operator)
...
+}
+
+TEST(TestS3SElect, null_arithmetic_operator)
+{
+  value a(7), d;
+  d.setnull();
+  ASSERT_EQ((a + d).to_string(), "nan" );
+  ASSERT_EQ((a - d).to_string(), "nan" );
+  ASSERT_EQ((a * d).to_string(), "nan" );
+  ASSERT_EQ((a / d).to_string(), "nan" );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:288]: need to add nan/0 (result is nan)

File: test/s3select_test.cpp [hunk 35]
```diff
@@ -222,6 +222,86 @@ TEST(TestS3SElect, arithmetic_operator)
...
+
+TEST(TestS3SElect, nan_arithmetic_operator)
+{
+  value a(7), d;
+  d.set_nan();
+  float b = ((a + d).dbl() );
+  float c = ((a - d).dbl() );
+  float v = ((a * d).dbl() );
+  float w = ((a / d).dbl() );
+  EXPECT_TRUE(!(b <= b) );
```
Comment (galsalomon66) [inline test/s3select_test.cpp:299]: that's good, it demonstrates both compare and arithmetic operators.
replace the /EXPECT_TRUE(!/ 
with
/EXPECT_FALSE( b<=b) /

it's more readable.

File: test/s3select_test.cpp [hunk 36]
```diff
@@ -629,3 +709,129 @@ TEST(TestS3selectFunctions, modzero)
...
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("null,\n"));
+} 
+
```
Comment (galsalomon66) [inline test/s3select_test.cpp:811]: please add also use-case such as  nullif(6,null)  ; nullif(5,"5") 
i.e. operands are of different types

File: include/s3select_oper.h [hunk 37]
```diff
@@ -776,12 +878,18 @@ class value
   {
     return compute<binop_mult>(*this, v);
   }
-
-  value& operator/(const value& v)  // TODO  handle division by zero
+  
+  value& operator/(value& v)  // TODO  handle division by zero
   {
-    return compute<binop_div>(*this, v);
+    if (v.type == value_En_t::S3NULL || this->type == value_En_t::S3NULL) {
```
Comment (galsalomon66) [inline include/s3select_oper.h:884]: please use is_null() instead of /type == value_En_t::S3NULL/

File: include/s3select_functions.h (hunk 31 above)
Comment (mattbenjamin) [reply include/s3select_functions.h:1008]: ew, ok

File: include/s3select_oper.h [hunk 38]
```diff
@@ -858,9 +883,15 @@ class value
     return compute<binop_mult>(*this, v);
   }
   
-  value& operator/(const value& v)  // TODO  handle division by zero
+  value& operator/(value& v)  // TODO  handle division by zero
   {
-    return compute<binop_div>(*this, v);
+    if (v.is_null() || this->is_null()) {
+      v.__val.dbl = NAN;
```
Comment (galsalomon66) [inline include/s3select_oper.h:889]: use set_nan()

File: test/s3select_test.cpp (hunk 35 above)
Comment (galsalomon66) [reply test/s3select_test.cpp:301]: replace with EXCEPT_FALSE

File: include/s3select_oper.h [hunk 39]
```diff
@@ -833,6 +852,12 @@ class value
         l.type = value_En_t::FLOAT;
       }
     }
+    if ((l.is_null() || r.is_null()) || (l.is_nan() || r.is_nan()))
+    {
+      l.__val.dbl = NAN;
```
Comment (galsalomon66) [inline include/s3select_oper.h:857]: better to use set_nan()

**Comment by galsalomon66** [comment]
wrong result for 
null+1 = 1  

the test is missing unit test for such case (it contain 1 + null)

**Comment by albin-antony** [comment]
Thank you Gal
